rskj-core/src/main/java/co/rsk/db/importer/StateImporter.java

```diff
@@ -0,0 +1,79 @@
+package co.rsk.db.importer;
+
+import co.rsk.RskContext;
+import co.rsk.trie.Trie;
+import co.rsk.trie.TrieStore;
+import org.ethereum.core.Block;
+import org.ethereum.db.BlockStore;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Iterator;
+import java.util.Objects;
+
+public class StateImporter {
+
+    private final BlockStore blockStore;
+    private final TrieStore trieStore;
+
+    private StateImporter(RskContext ctx) {
+        this.blockStore = ctx.getBlockStore();
+        this.trieStore = ctx.getTrieStore();
+    }
+
+    public static void importDataIfRequired(RskContext ctx) {
+        if (ctx.getRskSystemProperties().databaseReset() || !ctx.getRskSystemProperties().databaseImport()) {
+            return;
+        }
+
+        String databaseDir = ctx.getRskSystemProperties().databaseDir();
+
+        StateImporter importer = new StateImporter(ctx);
+        URL bootstrapURL = getBootstrapURL(ctx);
+        BootstrapDataProvider bootstrapDataProvider = new BootstrapDataProvider(databaseDir, bootstrapURL, ctx);
+        importer.updateDatabaseIfNeeded(bootstrapDataProvider.getBootstrapBlockstore(), bootstrapDataProvider.getBootstrapTrieStore());
+    }
+
+    private static URL getBootstrapURL(RskContext ctx) {
+        String url = ctx.getRskSystemProperties().databaseImportUrl();
+        try {
+            return new URL(url);
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException(e);
+        }
+    }
+
+    private void updateDatabaseIfNeeded(BlockStore bootstrapBlockStore, TrieStore bootstrapTrieStore) {
+        long maxBlockNumber = bootstrapBlockStore.getMaxNumber();
+        if (maxBlockNumber < blockStore.getMaxNumber()) {
+            return;
+        }
+
+        Block block = Objects.requireNonNull(bootstrapBlockStore.getChainBlockByNumber(maxBlockNumber));
+        updateBlockStore(bootstrapBlockStore, block);
+        updateTrieStore(bootstrapTrieStore, block);
+    }
+
+    private void updateBlockStore(BlockStore bootstrapBlockStore, Block bestBlock) {
+        Block current = bestBlock;
+        while (current != null) {
+            blockStore.saveBlock(current, bootstrapBlockStore.getTotalDifficultyForHash(current.getHash().getBytes()), true);
+            current = bootstrapBlockStore.getBlockByHash(current.getParentHash().getBytes());
+        }
+    }
+
+    private void updateTrieStore(TrieStore bootstrapTrieStore, Block bestBlock) {
+        Trie node = bootstrapTrieStore.retrieve(bestBlock.getStateRoot());
```

Comment:
You must take `StateRootHandler` into account

Reply:
Importing data from a previous height than wasabi won't be allowed. Thanks for appointing this :+1: